src/dynamic_program.rs

```diff
@@ -44,68 +84,205 @@ pub enum DynamicProgram {
     /// * Transaction::keys[0..] - program dependent
     /// * name - name of the program, translated to a file path of the program module
     /// * userdata - program specific user data
-    Native {
-        name: String,
-        library: libloading::Library,
-    },
+    Native { name: String, library: Library },
     /// Bpf program
     /// * Transaction::keys[0..] - program dependent
     /// * TODO BPF specific stuff
     /// * userdata - program specific user data
-    Bpf { userdata: Vec<u8> },
+    Bpf { name: String, prog: Vec<u8> },
 }
 
 impl DynamicProgram {
-    pub fn new(name: String) -> Self {
-        // TODO determine what kind of module to load
-
+    pub fn new_native(name: String) -> Self {
         // create native program
-        let path = create_library_path(&name);
+        let path = ProgramPath::Native {}.create(&name);
         // TODO linux tls bug can cause crash on dlclose, workaround by never unloading
-        let os_lib =
-            libloading::os::unix::Library::open(Some(path), libc::RTLD_NODELETE | libc::RTLD_NOW)
-                .unwrap();
-        let library = libloading::Library::from(os_lib);
+        let library = Library::open(Some(path), libc::RTLD_NODELETE | libc::RTLD_NOW).unwrap();
         DynamicProgram::Native { name, library }
     }
 
+    pub fn new_bpf_from_file(name: String) -> Self {
+        // create native program
+        let path = ProgramPath::Bpf {}.create(&name);
+        let file = match elf::File::open_path(&path) {
+            Ok(f) => f,
+            Err(e) => panic!("Error opening ELF {:?}: {:?}", path, e),
+        };
+
+        let text_section = match file.get_section(PLATFORM_SECTION_RS) {
+            Some(s) => s,
+            None => match file.get_section(PLATFORM_SECTION_C) {
+                Some(s) => s,
+                None => panic!("Failed to find text section"),
+            },
+        };
+        let prog = text_section.data.clone();
+
+        DynamicProgram::Bpf { name, prog }
+    }
+
+    pub fn new_bpf_from_buffer(prog: Vec<u8>) -> Self {
+        DynamicProgram::Bpf {
+            name: "from_buffer".to_string(),
+            prog,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn dump_prog(name: &str, prog: &[u8]) {
+        let mut eight_bytes: Vec<u8> = Vec::new();
+        println!("BPF Program: {}", name);
+        for i in prog.iter() {
+            if eight_bytes.len() >= 7 {
+                println!("{:02X?}", eight_bytes);
+                eight_bytes.clear();
+            } else {
+                eight_bytes.push(i.clone());
+            }
+        }
+    }
+
+    fn serialize(infos: &mut Vec<KeyedAccount>, data: &[u8]) -> Vec<u8> {
```

Comment:
Why serialize into a byte vector instead of defining a repr(C) struct to pass in/out of bpf vm?

Reply:
No strong pref right now (ie, let’s make it work first, optimize second).  The account userdata could be quite large so copying in/out of the VM is going to be not great in the long term (plus the account userdata has already been copied once by the bank before we even arrive here).  How much flexibility do we have with mapping host memory into the VM’s logical address space?

Reply:
No support right now in rbpf but something we can add